### ISSUE TITLE
How to fix Unhandled Cuda Error

### ISSUE THREAD

**Issue by light42** [facebookresearch/wav2letter#823]
### Question
I got this error

```
[[63962,1],0]: A high-performance Open MPI point-to-point messaging module
was unable to find any relevant network interfaces:

Module: OpenFabrics (openib)
  Host: dev-voice4

Another transport will be used instead, although this may result in
lower performance.

NOTE: You can disable this warning by setting the MCA parameter
btl_base_warn_component_unused to 0.
--------------------------------------------------------------------------
terminate called after throwing an instance of 'std::runtime_error'
  what():  unhandled cuda error
*** Aborted at 1599733111 (unix time) try "date -d @1599733111" if you are using GNU date ***
PC: @     0x7ff03490ff47 gsignal
*** SIGABRT (@0x3e800001023) received by PID 4131 (TID 0x7ff08f581f40) from PID 4131; stack trace: ***
    @     0x7ff035e7f8a0 (unknown)
    @     0x7ff03490ff47 gsignal
    @     0x7ff0349118b1 abort
    @     0x7ff058c07045 __gnu_cxx::__verbose_terminate_handler()
    @     0x7ff058b77276 __cxxabiv1::__terminate()
    @     0x7ff058b772c1 std::terminate()
    @     0x7ff058b6b023 __cxa_throw
    @     0x5618b1a91348 fl::detail::ncclCheck()
    @     0x5618b1a92a27 fl::distributedInit()
    @     0x5618b1a00007 w2l::initDistributed()
    @     0x5618b1773712 main
    @     0x7ff0348f2b97 __libc_start_main
    @     0x5618b17dcaaa _start
Aborted (core dumped)
```
when running

`../../../../../build/Train train --flagfile train_am_transformer_s2s_modified_nthread.cfg --minloglevel=0 --logtostderr=1
`

inside

`../wav2letter/recipes/models/sota/2019/librispeech
`

Here's the content of train_am_transformer_s2s_modified_nthread.cfg:
```

--runname=am_transformer_seq2seq_librispeech_6GPU_nthread
--rundir=/home/user/data_voice/run_model
--archdir=/home/user/wav2letter/wav2letter/recipes/models/sota/2019
--arch=am_arch/am_transformer_s2s.arch
--tokensdir=/home/user/data_voice/voice_sota_model/am
--tokens=librispeech-train-all-unigram-10000.tokens
--lexicon=/home/user/data_voice/voice_sota_model/am/librispeech-train+dev-unigram-10000-nbest10.lexicon
--train=/home/user/data/audio/lists/train-clean-100.lst,/home/user/data/audio/lists/train-clean-360.lst,/home/user/data/audio/lists/train-other-500.lst
--valid=dev-clean:/home/user/data/audio/lists/dev-clean.lst,dev-other:/home/user/data/audio/lists/dev-other.lst
--criterion=transformer
--mfsc
--usewordpiece=true
--wordseparator=_
--am_decoder_tr_dropout=0.1
--am_decoder_tr_layerdrop=0.1
--am_decoder_tr_layers=6
--maxdecoderoutputlen=120
--labelsmooth=0.05
--dataorder=output_spiral
--inputbinsize=25
--attnWindow=softPretrain
--softwstd=4
--trainWithWindow=true
--pretrainWindow=3
--attention=keyvalue
--encoderdim=256
--memstepsize=5000000
--eostoken=true
--pcttraineval=1
--pctteacherforcing=99
--sampletarget=0.01
--netoptim=adagrad
--critoptim=adagrad
--lr=0.03
--lrcrit=0.03
--linseg=0
--momentum=0.0
--maxgradnorm=0.1
--onorm=target
--sqnorm
--nthread=6
--batchsize=15
--filterbanks=80
--enable_distributed
--warmup=64000
--saug_start_update=64000
--lr_decay=200
--lr_decay_step=40
```

I don't understand the error message, so it's hard for me to pinpoint the mistake.

**Comment by tlikhomanenko** [comment]
Seems it is some error in NCCL initialization, could you make sure that flashlight tests are working for you (test on distributed run)?

**Comment by light42** [comment]
@tlikhomanenko Pardon my lack of understanding, but please give me step-by-step instruction to do flashlight tests.

**Comment by tlikhomanenko** [comment]
```
cd [PATH TO FL]/build
make test
```

**Comment by light42** [comment]
@tlikhomanenko it passed all the tests

```
Running tests...
Test project /home/user/flashlight/flashlight/build
      Start  1: AutogradTest
 1/18 Test  #1: AutogradTest .....................   Passed   49.26 sec
      Start  2: DevicePtrTest
 2/18 Test  #2: DevicePtrTest ....................   Passed   10.68 sec
      Start  3: HistogramTest
 3/18 Test  #3: HistogramTest ....................   Passed   11.47 sec
      Start  4: LoggingTest
 4/18 Test  #4: LoggingTest ......................   Passed    9.52 sec
      Start  5: SerializationTest
 5/18 Test  #5: SerializationTest ................   Passed    9.37 sec
      Start  6: OptimTest
 6/18 Test  #6: OptimTest ........................   Passed   10.65 sec
      Start  7: CachingMemoryManagerTest
 7/18 Test  #7: CachingMemoryManagerTest .........   Passed   11.04 sec
      Start  8: MemoryFrameworkTest
 8/18 Test  #8: MemoryFrameworkTest ..............   Passed   10.93 sec
      Start  9: MemoryInitTest
 9/18 Test  #9: MemoryInitTest ...................   Passed    9.99 sec
      Start 10: ModuleTest
10/18 Test #10: ModuleTest .......................   Passed   12.39 sec
      Start 11: NNSerializationTest
11/18 Test #11: NNSerializationTest ..............   Passed   13.11 sec
      Start 12: NNUtilsTest
12/18 Test #12: NNUtilsTest ......................   Passed   10.64 sec
      Start 13: DatasetTest
13/18 Test #13: DatasetTest ......................   Passed   10.80 sec
      Start 14: DatasetUtilsTest
14/18 Test #14: DatasetUtilsTest .................   Passed    9.33 sec
      Start 15: MeterTest
15/18 Test #15: MeterTest ........................   Passed   10.77 sec
      Start 16: AllReduceTest
16/18 Test #16: AllReduceTest ....................   Passed    9.90 sec
      Start 17: ContribModuleTest
17/18 Test #17: ContribModuleTest ................   Passed   12.53 sec
      Start 18: ContribSerializationTest
18/18 Test #18: ContribSerializationTest .........   Passed   12.12 sec

100% tests passed, 0 tests failed out of 18

Total Test time (real) = 234.57 sec
```

**Comment by tlikhomanenko** [comment]
Ok, can you remove `--enable_distributed` and run training (this will run on 1 GPU). If this works, you need to correctly start training on several gpus. Either with mpi and providing --enable_distributed, or with some other distributed tools like SLURM (here you need to specify rndz, worldrank and worldsize flags). 

FYI: n_threads is not number of gpus to use.

**Comment by light42** [comment]
> Ok, can you remove `--enable_distributed` and run training (this will run on 1 GPU). If this works, you need to correctly start training on several gpus. Either with mpi and providing --enable_distributed, or with some other distributed tools like SLURM (here you need to specify rndz, worldrank and worldsize flags).
> 
> FYI: n_threads is not number of gpus to use.

Yes, this does the trick, eventhough it made the program can't run at parallel but I'm fine with that.

Actually I'm not sure what those flags mean, because I can't find the documentation for all of them, and I'm not savvy enough to know where exactly those flags would appear on source code.

nthread flag is not necessary but I'm adding it to find out whether it could improve training speed. And also I don't know which flags to use to set the number of training epochs, because wav2letter keep stopping at 2nd epoch eventhough I already set --epoch=10.

In conclusion, all of this happens because I can't find the documentation for the flags.

**Comment by tlikhomanenko** [comment]
Thanks for feedback! We are preparing huge documentation, but it will be available only in several weeks after. 

For now here is the list of flags https://github.com/facebookresearch/wav2letter/blob/v0.2/src/common/Defines.cpp#L20

To set number of epochs use `--iter` which is the number of updates (not epochs), so you can just set `--iter=1000000` some huge value, so training will go on.

Here is example how to run on 1 node with mpi https://github.com/facebookresearch/wav2letter/wiki/Distributed-training, and a bit of flags are described here https://github.com/facebookresearch/wav2letter/wiki/Train-a-model.

**Comment by light42** [comment]
@tlikhomanenko I'm sorry that I'm replying to closed and unrelated thread. I just want to ask, which flag I could set to improve training speed?

Through several experiment, I found that nthread and nthread_decoder had no effect whatsoever towards training speed.

**Comment by tlikhomanenko** [comment]
nthread_decoder is for decoding to run in beam-search decoding with data in parallel. nthread is about reading the dataset.

What do you mean by "improve training speed"?

**Comment by light42** [comment]
I mean, what flags I could probably set to shorten training time. 

In the log I found that one epoch usually finished around 45 minutes. 

I want to push this further until one epoch could be finished in 35 minutes.

**Comment by tlikhomanenko** [comment]
There are the following options:
- increase batchsize
- use more gpus
- use less data
- decrease model size (less transformer layers for example in the arch file)
- decrease number of layers in the AM-decoder for s2s `--am_decoder_tr_layers`

Entire operations are already highly optimized themselves.
